File: code/Project_4/BT5/c++/exercises/exercise_1_1.cpp
#include <iostream>
#include <fstream>
#include <set>
#include <string>
#include <sstream>
#include <algorithm>
using namespace std;

class Graph {
private:
    set<int> vertices;
    set<pair<int, int>> edges;

public:
    void add_edge(int u, int v) {
        vertices.insert(u);
        vertices.insert(v);
        edges.insert({min(u, v), max(u, v)});
    }
    
    string get_dimacs_format() {
        stringstream ss;
        ss << "p edge " << vertices.size() << " " << edges.size() << "\n";
        for (const auto& edge : edges) {
            ss << "e " << edge.first << " " << edge.second << "\n";
        }
        return ss.str();
    }
    
    static Graph read_dimacs_file(const string& filename) {
        Graph graph;
        ifstream file(filename);
        if (!file.is_open()) {
            cout << "File " << filename << " not found\n";
            return graph;
        }
        
        string line;
        while (getline(file, line)) {
            if (line.empty()) continue;
            
            if (line[0] == 'c') {
                // Comment line - bỏ qua
                continue;
            } else if (line[0] == 'p') {
                // Problem line
                continue;
            } else if (line[0] == 'e') {
                // Edge line
                stringstream ss(line);
                string type;
                int u, v;
                ss >> type >> u >> v;
                graph.add_edge(u, v);
            }
        }
        return graph;
    }
};

int main() {
    // Tạo đồ thị ví dụ
    Graph g;
    g.add_edge(1, 2);
    g.add_edge(2, 3);
    g.add_edge(1, 3);
    g.add_edge(3, 4);
    
    // Ghi ra định dạng DIMACS
    cout << "DIMACS format:\n";
    cout << g.get_dimacs_format();
    
    // Ghi vào file
    ofstream outfile("sample.dimacs");
    outfile << g.get_dimacs_format();
    outfile.close();
    
    // Đọc lại từ file
    cout << "\nReading from file:\n";
    Graph g2 = Graph::read_dimacs_file("sample.dimacs");
    cout << g2.get_dimacs_format();
    
    return 0;
}

File: code/Project_4/BT5/python/exercises/exercise_1_1.py
class Graph:
    def __init__(self):
        self.vertices = set()
        self.edges = set()
    
    def add_edge(self, u, v):
        self.vertices.add(u)
        self.vertices.add(v)
        self.edges.add((min(u, v), max(u, v)))  # Undirected graph
    
    def get_dimacs_format(self):
        """Chuyển đồ thị sang định dạng DIMACS"""
        lines = []
        lines.append(f"p edge {len(self.vertices)} {len(self.edges)}")
        for u, v in sorted(self.edges):
            lines.append(f"e {u} {v}")
        return "\n".join(lines)
    
    @staticmethod
    def read_dimacs_file(filename):
        """Đọc đồ thị từ file DIMACS"""
        graph = Graph()
        try:
            with open(filename, 'r') as f:
                for line in f:
                    line = line.strip()
                    if line.startswith('c'):
                        # Comment line - bỏ qua
                        continue
                    elif line.startswith('p'):
                        # Problem line - p edge n m
                        parts = line.split()
                        if len(parts) >= 4:
                            n, m = int(parts[2]), int(parts[3])
                    elif line.startswith('e'):
                        # Edge line - e i j
                        parts = line.split()
                        if len(parts) >= 3:
                            u, v = int(parts[1]), int(parts[2])
                            graph.add_edge(u, v)
        except FileNotFoundError:
            print(f"File {filename} not found")
        return graph

def main():
    # Tạo đồ thị ví dụ
    g = Graph()
    g.add_edge(1, 2)
    g.add_edge(2, 3)
    g.add_edge(1, 3)
    g.add_edge(3, 4)
    
    # Ghi ra định dạng DIMACS
    print("DIMACS format:")
    print(g.get_dimacs_format())
    
    # Ghi vào file
    with open('sample.dimacs', 'w') as f:
        f.write(g.get_dimacs_format())
    
    # Đọc lại từ file
    print("\nReading from file:")
    g2 = Graph.read_dimacs_file('sample.dimacs')
    print(g2.get_dimacs_format())

if __name__ == "__main__":
    main()